prov/efa/src/efa_av.c

```diff
@@ -342,26 +356,30 @@ int efa_rdm_av_insert_addr(struct efa_av *av, struct efa_ep_addr *addr,
 			EFA_WARN(FI_LOG_AV,
 				 "Max number of shm AV entry %d has been reached.\n",
 				 rxr_env.shm_av_size);
-			goto err_free_av_entry;
+			ofi_av_remove_addr(&av->util_av, *fi_addr);
+			goto out;
+
 		}
 		ret = rxr_ep_efa_addr_to_str(addr, smr_name);
-		if (ret != FI_SUCCESS)
-			goto err_free_av_entry;
+		if (ret != FI_SUCCESS) {
+			EFA_WARN(FI_LOG_AV,
+				 "rxr_ep_efa_addr_to_str() failed! ret=%d\n", ret);
+			ofi_av_remove_addr(&av->util_av, *fi_addr);
+			goto out;
+		}
 
-		ret = fi_av_insert(av->shm_rdm_av, smr_name, 1, &shm_fiaddr,
-					flags, context);
+		ret = fi_av_insert(av->shm_rdm_av, smr_name, 1, &shm_fiaddr, flags, context);
 		if (OFI_UNLIKELY(ret != 1)) {
 			EFA_WARN(FI_LOG_AV,
 				 "Failed to insert address to shm provider's av: %s\n",
 				 fi_strerror(-ret));
-			goto err_free_av_entry;
-		} else {
-			ret = 0;
```

Comment:
It might be clearer if there were a pound-defined FI_AV_INSERT_SUCCESS 1 ... unless there are lots of other cases where there is a non-zero return code that indicates SUCCESS and there is no pound-defined definition to go with it.